src/library.cpp

```diff
@@ -157,6 +157,90 @@ bool Library::removeBookmark(const std::string& zimId, const std::string& url)
   return false;
 }
 
+int Library::migrateBookmarks() {
+  std::set<std::string> sourceBooks;
+  {
+    std::lock_guard<std::recursive_mutex> lock(m_mutex);
+    for(auto& bookmark:m_bookmarks) {
+      if (m_books.find(bookmark.getBookId()) == m_books.end()) {
+        sourceBooks.insert(bookmark.getBookId());
+      }
+    }
+  }
+  int changed = 0;
+  for(auto& sourceBook:sourceBooks) {
+    changed += migrateBookmarks(sourceBook);
+  }
+  return changed;
+}
+
+std::string Library::getBestTargetBookId(const Bookmark& bookmark) const {
+  // Search for a existing book with the same name
+  auto book_filter = Filter();
+  if (!bookmark.getBookName().empty()) {
+    book_filter.name(bookmark.getBookName());
+  } else {
+    // We don't have a name stored (older bookmarks)
+    // Fallback on title (All bookmarks should have one, but let's be safe against wrongly filled bookmark)
+    if (bookmark.getBookTitle().empty()) {
+        // No bookName nor bookTitle, no way to find target book.
+        return "";
+    }
+    book_filter.query("title:\""+bookmark.getBookTitle() + "\"");
+  }
+  auto targetBooks = filter(book_filter);
+  // Remove source book
+  auto foundIT = std::find(targetBooks.begin(), targetBooks.end(), bookmark.getBookId());
+  if (foundIT != targetBooks.end()) {
+    targetBooks.erase(foundIT);
+  }
+    
+  if (targetBooks.empty()) {
+    // No existing book found for the target, or the same than source.
+    return "";
+  }
+  if (targetBooks.size() != 1) {
+    sort(targetBooks, DATE, false);
+  }
+  return targetBooks[0];
+}
+
+int Library::migrateBookmarks(const std::string& source) {
+  std::lock_guard<std::recursive_mutex> lock(m_mutex);
+
+  Bookmark firstBookmarkToChange;
+  for(auto& bookmark:m_bookmarks) {
+    if (bookmark.getBookId() == source) {
+      firstBookmarkToChange = bookmark;
+      break;
+    }
+  }
+
+  if (firstBookmarkToChange.getBookId().empty()) {
+    return 0;
+  }
+
+  std::string betterBook = getBestTargetBookId(firstBookmarkToChange);
+
+  if (betterBook.empty()) {
+    return 0;
+  }
+
+  return migrateBookmarks(source, betterBook);
+}
+
+int Library::migrateBookmarks(const std::string& source, const std::string& destination) {
+  // We finally a set of bookmarks to change and a destination
+  int changed = 0;
+  for (auto& bookmark:m_bookmarks) {
+    if (bookmark.getBookId() == source) {
+      bookmark.setBookId(destination);
+      changed +=1;
+    }
```

Comment:
Shouldn't there be any foolproofness here? I guess a bookmark should be allowed to migrate only if the bookmarked path is present in the destination book.

Reply:
We agreed with @kelson42 to not check for the path but only for bookId